Remove 'Document' from CompletionItem

CyrusNajmabadi opened the pull request:
As part of teh "show completion for items you don't have usings for" feature, we'd like to be able to cache CompletionItems so we don't have to keep recreating vast numbers of the same items over and over again.

This is slightly problematic because CompletionItems today hold onto `Document`s, and that means an entire snapshot of the world from teh past is held onto.  

This PR changes CompletionItem to be pure data, without a hard link to a potentially very expensive piece of data like `Document`.

CyrusNajmabadi commented:
Tagging @ivanbasov @jasonmalinowski @genlu

CyrusNajmabadi commented on src/EditorFeatures/Core.Wpf/Completion/Presentation/CompletionPresenterSession.cs:
```diff
@@ -132,6 +135,7 @@ internal sealed class CompletionPresenterSession : ForegroundThreadAffinitizedOb
                 // so that we can call back into ourselves, get the items and add it to the
                 // session.
                 _editorSessionOpt.Properties.AddProperty(Key, this);
+                _editorSessionOpt.Properties.AddProperty(DocumentKey, document);
```
for the old, synchronous, completoin system, we just store the Document on the ephemeral ICompletionSession.  Note: we don't need to do this for the new, async, completion service.  In that world, the IAsyncCompletionSession has kept track of the trigger snapshot, and we use that to get back to the Document in question.

genlu commented on src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller.cs:
```diff
@@ -151,7 +151,7 @@ internal override void OnModelUpdated(Model modelOpt)
                 // All items from the completion list have .Document set to current document 
```
Comment needs to be updated.

jasonmalinowski commented on src/EditorFeatures/Core.Wpf/Completion/Presentation/CompletionPresenterSession.cs:
```diff
@@ -20,6 +20,8 @@ namespace Microsoft.CodeAnalysis.Editor.Implementation.IntelliSense.Completion.P
 
     internal sealed class CompletionPresenterSession : ForegroundThreadAffinitizedObject, ICompletionPresenterSession
     {
+        public static readonly object DocumentKey = new object();
```
Just add a comment what this is used for (i.e. it's a key added to which type?)

CyrusNajmabadi replied:
done!

jasonmalinowski commented on src/EditorFeatures/Core.Wpf/Completion/Presentation/CustomCommitCompletion.cs:
```diff
@@ -58,17 +57,12 @@ public void Commit()
-                service.GetDescriptionAsync(this.CompletionItem.Document, this.CompletionItem, cancellationToken);
-        }
-
-        public string GetDescription_TestingOnly()
```
Good to see this disappear!

jasonmalinowski commented on src/EditorFeatures/Core/Extensibility/Completion/ICompletionPresenterSession.cs:
```diff
@@ -14,7 +14,7 @@ internal interface ICompletionPresenterSession : IIntelliSensePresenterSession
             CompletionItem suggestionModeItem, bool suggestionMode, bool isSoftSelected,
             ImmutableArray<CompletionItemFilter> completionItemFilters,
-            string filterText);
+            string filterText, Document document);
```
FYI to @KirillOsenkov that this will impact Xamarin next time they take an update. Impact will be trivial, but just giving him a heads up.

jasonmalinowski commented:
@genlu @CyrusNajmabadi If @genlu is still planning on taking his feature into Preview 2, this would *also* have to go into Preview 2. Right now this branch is targeting after that. There's plenty of ways you can deal with this (move this back, cherry-pick this into his PR, etc.) so it doesn't matter to me. I would say if you can avoid merging *both* PRs into different branches that's probably best, only because you'll potentially create a merge mess for our swat team.

CyrusNajmabadi commented:
@jasonmalinowski I've rebased.  I also moved to just storing an ITextSnapshot instead of a Document.  That's what the IAsyncCompletionSession does, so i figured i'd just match that.  That way, if there somehow is a leak of the session, the worst case is that we're just holding onto an editor snapshot, not a full Roslyn solution snapshot.

genlu commented:
@CyrusNajmabadi do you want to merge this in to my PR instead and bring it for approval as a pacakge?

CyrusNajmabadi commented:
> @CyrusNajmabadi do you want to merge this in to my PR instead and bring it for approval as a pacakge?

I honestly don't care :) If you guys want to take it as an independent PR, great.  If you want to merge into your PR, that's fine with me too :)

genlu commented:
@CyrusNajmabadi OK, I will merge it into my PR. Then we can better test the whole feature :)
Thanks!

jasonmalinowski commented:
> @CyrusNajmabadi: I also moved to just storing an ITextSnapshot instead of a Document. That's what the IAsyncCompletionSession does, so i figured i'd just match that.

Do you have any concerns for perf? It means any time we do have to grab a document we're going to re-parse the _old_ text, and fork a Compilation all over again. It's a standard CPU vs. memory tradeoff; do we think it makes sense?

CyrusNajmabadi commented:
> Do you have any concerns for perf? It means any time we do have to grab a document we're going to re-parse the old text, and fork a Compilation all over again. It's a standard CPU vs. memory tradeoff; do we think it makes sense?

yes, def a concern.  but it's also something that seems to work fine locally inside roslyn.sln :)   and it's what IAsyncCompletion does... so i feel somewaht ok with this being acceptable :)

CyrusNajmabadi commented:
@ivanbasov @genlu @dpoeschl @jasonmalinowski can you review https://github.com/dotnet/roslyn/pull/34952/commits/430865b87cf71303bfd9821d45bbe1e94c0d0d7e as well?

jasonmalinowski commented:
@CyrusNajmabadi Well, it looked good either way. :smile:

genlu commented:
We can close this now, all changed merged in #34808
